### PR TITLE
Improve label format enforcement

### PR DESCRIPTION
**Pull request by fspmarshall**
Label key format requirements were only being enforced by json schema, which is only run on a subset of *deserializing* operations, and (unsurprisingly) only when the the format of origin is json.  As a result, invalid labels were not detected when passing through the GRPC API, being inserted into the backend, or when propagating through the event system.  This resulted in teleport working mostly correctly up until the point where it is *restarted*, at which point initialization of the auth server cache would fail, causing startup to fail.

This PR addresses the specific issue of labels, but I'm working on drafting an issue about the broader problem of inconsistent validation.  I'm generally of the opinion that json schema validation in teleport does more harm than good, and I think problems like this are a prime example of why validation logic should not be tied to a specific serialization format.

Partially addresses #4034

### CODE REVIEW

**Comment by fspmarshall** [comment]
retest this please

**Review by webvictim — COMMENTED** [review]
LGTM - wait for another review though!

File: lib/srv/regular/sshserver.go
Comment (awly) [inline lib/srv/regular/sshserver.go:511]: Move all these validations into the `ServerOption`s that set them.

File: lib/services/resource.go
Comment (awly) [inline lib/services/resource.go:930]: Add a unit test for `IsValidLabelKey` to make sure the regexp is correct

Comment (russjones) [reply lib/services/resource.go:930]: I would define the pattern and the `*regexp.Regexp` separately so you can use the same string in `CheckAndSetDefaults` and the JSON schema to avoid drift. Something like:

```go
const labelPattern = `^[a-zA-Z/.0-9_*-]+$`
var validLabelKey = regexp.MustCompile(labelPattern)
```

```go
const MetadataSchema = `{
  "type": "object",
  "additionalProperties": false,
  "default": {},
  "required": ["name"],
  "properties": {
    "name": {"type": "string"},
    "namespace": {"type": "string", "default": "default"},
    "description": {"type": "string"},
    "expires": {"type": "string"},
    "id": {"type": "integer"},
    "labels": {
      "type": "object",
      "additionalProperties": false,
      "patternProperties": {
         "%v":  { "type": "string" }
      }
    }
  }
}`
```

Then you will probably have to update the the functions that return schemas to do something like:

```go
return fmt.Sprintf(V3SchemaTemplate, MetadataSchema, labelPattern, [...])

**Review by russjones — APPROVED** [review]
Looks good to me once the issues in the comments are addressed.

File: lib/srv/regular/sshserver.go
Comment (awly) [inline lib/srv/regular/sshserver.go:354]: Combine this with the loop below.
Also, should `labels` be cloned to prevent concurrent writes?

**Review by awly — APPROVED** [review]

File: lib/srv/regular/sshserver.go
Comment (fspmarshall) [reply lib/srv/regular/sshserver.go]: As far as I can tell, only `cmdLabels` is mutated.  The `labels` value should be calculated once at program start based on the config file and/or command-line inputs.

**Pull request merged**
